### PR TITLE
[ENH] Added TCN forecaster in aeon/forecasting/deep_learning

### LINKED ISSUE #2937: [ENH]  Implement the TCNForecaster in `aeon/forecasting/deep_learning`
### Describe the feature or idea you want to propose

I propose the addition of a new forecaster class, `TCNForecaster`, under the module `aeon.forecasting.deep_learning`. This forecaster will use a Temporal Convolutional Network (TCN) for time series forecasting, built upon aeon's `BaseDeepForecaster` and a modular `TCNNetwork` class.
Refer #2931 

### Describe your proposed solution

The proposed implementation introduces a `TCNForecaster` class which:

- Inherits from `BaseDeepForecaster`
- Uses a configurable `TCNNetwork` to construct the underlying architecture
- Supports multi-variate input, customizable kernel size, dropout, and block depths

The forecaster is structured to follow aeon's deep learning forecaster pattern, enabling compatibility with the overall API and model selection tools.

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lucifer4073**
#### Reference Issues/PRs

Fixes #2937.

#### What does this implement/fix? Explain your changes.
Adds TCNForecaster in aeon/toolkit.

##### For all contributions
- [X] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you **after** the PR has been merged.
- [X] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [X] I've added the estimator/function to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [ ] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

File: aeon/forecasting/deep_learning/_tcn.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+from aeon.forecasting.base import DirectForecastingMixin
+from aeon.forecasting.deep_learning.base import BaseDeepForecaster
+from aeon.networks._tcn import TCNNetwork
+
+
+class TCNForecaster(BaseDeepForecaster, DirectForecastingMixin):
+    """A deep learning forecaster using Temporal Convolutional Network (TCN).
+
+    It leverages the `TCNNetwork` from aeon's network module
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/_tcn.py]: parameters missing are:

save_last_model, save_init_model and best_model_name and init_model_name, please check any example in aeon/classification/deep_learning to see how to use  them

File: aeon/forecasting/deep_learning/_tcn.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+        input_shape : tuple
+            Shape of input data, typically (window, num_inputs).
+
+        Returns
+        -------
+        model : tf.keras.Model
+            Compiled Keras model with TCN architecture.
+        """
+        import tensorflow as tf
+
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/_tcn.py:163]: the set of random state stuff should be done here in build_model not in fit

File: aeon/forecasting/deep_learning/_tcn.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        network = TCNNetwork(
+            n_blocks=self.n_blocks if self.n_blocks is not None else [16, 16, 16],
+            kernel_size=self.kernel_size,
+            dropout=self.dropout,
+        )
+        input_layer, output = network.build_network(input_shape=input_shape)
+        model = tf.keras.Model(inputs=input_layer, outputs=output)
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/_tcn.py:176]: after creating the model in build_model you should do the following:

```
self.optimizer_ = (
            tf.keras.optimizers.Adam() if self.optimizer is None else self.optimizer
        )
```
and then compile the model before returning. Compiling should not be done in fit

File: aeon/forecasting/deep_learning/_tcn.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    }
+
+    def __init__(
+        self,
+        horizon=1,
+        window=10,
+        batch_size=32,
+        n_epochs=100,
+        verbose=0,
+        optimizer="adam",
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/_tcn.py]: optimizer should be default to None as the input should be a keras optimizer not a string

File: aeon/forecasting/deep_learning/_tcn.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        network = TCNNetwork(
+            n_blocks=self.n_blocks if self.n_blocks is not None else [16, 16, 16],
+            kernel_size=self.kernel_size,
+            dropout=self.dropout,
+        )
+        input_layer, output = network.build_network(input_shape=input_shape)
+        model = tf.keras.Model(inputs=input_layer, outputs=output)
+        return model
+
+    def _fit(self, y, exog=None):
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/_tcn.py:190]: Please take a look at this example from deep classification in aeon to see how to handle saving best last and init model, and their names, please do the exact same thing, especially the loading of the best model at the end where we do try and except:

https://github.com/aeon-toolkit/aeon/blob/9357e2fac7579f0fd058e4dca345e22f11dcb971/aeon/classification/deep_learning/_fcn.py#L231

File: aeon/forecasting/deep_learning/_tcn.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+
+import numpy as np
+from sklearn.utils import check_random_state
+
+from aeon.forecasting.base import DirectForecastingMixin
+from aeon.forecasting.deep_learning.base import BaseDeepForecaster
+from aeon.networks._tcn import TCNNetwork
+
+
+class TCNForecaster(BaseDeepForecaster, DirectForecastingMixin):
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/_tcn.py]: TCN should inherit mixin of iterartive not direct. As it predicts one value, and then the ```iterative_forecast``` function should be re-implemented for TCNForecaster because by default it doesn handle multivariate but TCN does so it should be re-implemented

File: aeon/forecasting/deep_learning/_tcn.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+    file_path : str, default="./"
+        Directory path where models will be saved.
+    n_blocks : list of int, default=[16, 16, 16]
+        List specifying the number of output channels for each layer of the
+        TCN. The length determines the depth of the network.
+    kernel_size : int, default=2
+        Size of the convolutional kernel in the TCN.
+    dropout : float, default=0.2
+        Dropout rate applied after each convolutional layer for
+        regularization.
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/_tcn.py]: Add reference paper for TCN and cite it in the docstring

**Comment by lucifer4073** [comment]
Hi @hadifawaz1999 ,

I have made the requested changes.

File: aeon/forecasting/deep_learning/_tcn.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+import numpy as np
+from sklearn.utils import check_random_state
+
+from aeon.forecasting.base import IterativeForecastingMixin
+from aeon.forecasting.deep_learning.base import BaseDeepForecaster
+from aeon.networks._tcn import TCNNetwork
+
+
+class TCNForecaster(BaseDeepForecaster, IterativeForecastingMixin):
+    """A deep learning forecaster using Temporal Convolutional Network (TCN).
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/_tcn.py:23]: cite the reference here [1]_

File: aeon/forecasting/deep_learning/tests/test_tcn.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    forecaster = TCNForecaster(
+        horizon=horizon, window=window, n_epochs=epochs, batch_size=16, verbose=0
+    )
+
+    # Fit and predict
+    forecaster.fit(y)
+    prediction = forecaster.predict(y)
+
+    # Basic assertions
+    assert prediction is not None
```
Comment (hadifawaz1999) [inline aeon/forecasting/deep_learning/tests/test_tcn.py]: can you test the shape of the output of prediction ?

Comment (lucifer4073) [reply aeon/forecasting/deep_learning/tests/test_tcn.py]: The prediction is a single `float` value. I can't use shape/len so instead I can check the type.

**Review by hadifawaz1999 — APPROVED** [review]
